### ISSUE TITLE
EXPORT_ES6: Add testing and support for node.js (replace require(), __dirname)

### ISSUE THREAD

**Issue by curiousdannii** [emscripten-core/emscripten#11792]
So I'm finally getting around to trying EXPORT_ES6 mode, and it just doesn't work in node (when loaded as an actual ES6 Module), because it still refers to `__dirname`. I assume other people have only been using it Webpack/rollup/whatever?

It doesn't look like there are any tests for EXPORT_ES6 either.

**Comment by kripken** [comment]
I think we don't have tests for it because at the time there wasn't a version of node that supported it. If there is now, we should add some.

I think we have 12.10 on CI, and can maybe add another version. Another option is to use v8, which we already install a new version of, if it has ES6 module support.

It would be good if someone could write up a test for this and then we can check the options.

**Comment by curiousdannii** [comment]
Node 12.10 will definitely be enough.

**Comment by sbc100** [comment]
Just FYI, it looks like out current version of node that we use in emsdk and in all our testing is `12.18.1`.

**Comment by jvail** [comment]
Could this issue be extended to cover also "require" vs "import"? I wonder if replacing all lines in the output such as 

```js
var nodeFS = require('fs');`
```

with 

```js
#if EXPORT_ES6
const { default: nodeFS } = await import('fs');
#else
var nodeFS = require('fs');
```
would be sufficient.

Unfortunately my insight in the build system is very limited. But if someone could start then I could help at least with testing.

**Comment by curiousdannii** [comment]
Realised this is going to be difficult to handle for multi-environment builds. ES import statements can't be put in conditionals, and while the import function can, it needs to be handled async.

**Comment by RReverser** [comment]
> ES import statements can't be put in conditionals, and while the import function can, it needs to be handled async.

They don't need to, Emscripten could emit static import declarations instead.

**Comment by curiousdannii** [comment]
I said it would be difficult for multi-environment builds. A static import declaration for node will fail in the browser.

**Comment by RReverser** [comment]
Ah, right, I see. In that case, the async import sounds fine. The fact that it's async shouldn't be an issue - Emscripten already has a set of async runtime dependencies and a way to wait on them.

Take a look at usages of `addRunDependency` from https://github.com/emscripten-core/emscripten/blob/0cc4a667828be0793db01c8c6b85434b6ccb38dc/src/preamble.js#L488.

**Comment by matthewp** [comment]
@RReverser You mentioned working on this in #13571 , do you have a test case that you were using?

**Comment by RReverser** [comment]
I said I stopped working on it and just waiting for someone else to fix it :)

Absolutely any file in `EXPORT_ES6`. Try to generate `emcc temp.c -o temp.mjs` (it will export ES6 module) and then

```js
import Module from './temp.mjs';
Module();
```

somewhere else. You'll run into the code trying to use `__dirname`.

The fix for just one configuration wouldn't be hard, but the general issue is that `scriptDirectory` is assigned from multiple places under different modes, and there needs to be some more systematic fix to align them correctly.

**Comment by lovasoa** [comment]
What is the best workaround for this at the moment ? Not using EXPORT_ES6  ?

**Comment by curiousdannii** [comment]
Yes, just don't use EXPORT_ES6. If you're targeting Node that won't be a problem, even if the rest of your app uses ES modules you'll be able to make it work (possibly by changing the file extension to .cjs, I can't remember the details.)

**Comment by matthewp** [comment]
@lovasoa I still use `EXPORT_ES6`, I just have a post-build script that removes the commonjs-isms. I do wind up producing separate Node.js and Browser entry points as a result though.

**Comment by yhung119** [comment]
@matthewp I'm using `EXPORT_ES6` as well, but I haven't figure out a way to get it compiled. I am able to get pass the `__dirname` error, but then it complained about `require`.
I'm curious how you did it. Can you point me to resources you used or the github repo with a working sample? I'm fairly new to this. many thanks in advanced :)

**Comment by surma** [comment]
In case it’s helpful, this has been my workaround for making `EXPORT_ES6` files work in Node.js:

```js
import {dirname} from "path";
globalThis.__dirname = dirname(import.meta.url);
import { createRequire } from 'module';
globalThis.require = createRequire(import.meta.url);

import emscriptenModule from "./my-emscripten-module.js";
```

**Comment by matthewp** [comment]
@yhung119 I just use sed to find and replace the require calls with references to `path`, etc. and then append appropriate import statements to the top. It does mean that I have separate JS files for Node and Browser which isn't great. Perhaps there is a better way. You can see what I'm doing here: https://github.com/lucydsl/liblucy/blob/77fb82fa6c72a3621de60da61c9d972da0659e6e/Makefile#L19-L26

**Comment by thomasballinger** [comment]
Here's smaller example of roughly the workaround described by @matthewp above along with a minimal repro (`make node-repro`) of the issue. https://github.com/thomasballinger/embind-ES6-for-node-and-browser

**Comment by kripken** [comment]
It looks like this is the best we have to a tracking issue for the general problems with the current `EXPORT_ES6`. See also the linked issues, and also some relevant discussion in https://github.com/WebAssembly/binaryen/pull/3304#issuecomment-721322208

Following that, some offline discussion raised the following ideas:
* For proper ES6 module support, we don't want `MODULARIZE` mode - we don't need a factory. We just want to emit the code without a factory, directly - being emitted in an ES6 module file takes care of the code being "modular".
* It seems hard to change the current relevant flags without breaking changes, `EXPORT_ES6` and `MODULARIZE`, and we used to have `MODULARIZE_INSTANCE`.
* Instead, perhaps we can deprecate `EXPORT_ES6`, and add a new flag `ESM` (or `ES_MODULE` or such). The goal would be for that to emit exactly the right contents for an ES6 module: no factory, the exports are the expected ones, etc.
* When that is stable, we could remove the deprecated `EXPORT_ES6`.

Thoughts?

**Comment by RReverser** [comment]
> It looks like this is the best we have to a tracking issue for the general problems with the current EXPORT_ES6.

Hmm, I wouldn't mix them together tbh, but instead raise a new issue if there isn't one for ES6 already. This issue is mostly about Node.js support which has its own ES6+Node.js specific challenges regardless of chosen output format.

**Comment by sbc100** [comment]
> > It looks like this is the best we have to a tracking issue for the general problems with the current EXPORT_ES6.
> 
> Hmm, I wouldn't mix them together tbh, but instead raise a new issue if there isn't one for ES6 already. This issue is mostly about Node.js support which has its own ES6+Node.js specific challenges regardless of chosen output format.

I agree we can/should probably make a seperate issue.   I am a little confused by the above statement thought.  Isn't this issue all about module?   The use of require() and __dirname are perfect fine under normal non-module usage aren't they?

**Comment by RReverser** [comment]
>  Isn't this issue all about module? The use of require() and __dirname are perfect fine under normal non-module usage aren't they?

Right, but what I'm saying is those issues will happen both in factory output and in output with individual exports, so those different output modes are not really relevant to this Node.js-specific issue.

**Comment by curiousdannii** [comment]
Factory mode is still important! Why do you think it isn't desired anymore?

**Comment by kripken** [comment]
@curiousdannii 

I should have said that factory mode should be optional. Enabling `MODULARIZE` could still create the factory function as always.

**Comment by curiousdannii** [comment]
In #9325 I suggested

> What if we deprecated MODULARIZE in favour of a new setting named something like OUTPUT_MODE with options 'plain', 'factory', 'instance'

Likewise we could have a `MODULE_FORMAT` setting with options 'none', 'esm', 'umd'. Perhaps the 'none' and 'umd' options could automatically wrap in an IIFE?

These are tangential options, so an option that mixes them like your proposed `ESM` option makes less sense to me. It's just one out of the 9 potential combinations. And if we change the default to 'esm', then it will be even less necessary.

**Comment by kripken** [comment]
Oh, good points. I guess I'm not sure what the best options here would be. Perhaps we could rename `MODULARIZE` to `FACTORY` as you said. Then specifying `EXPORT_ES6` with or without `FACTORY` would indicate the mode. But if umd is important than a new setting might make sense - I'm not close enough to the JS ecosystem to know what's best here.

I haven't had time for #12203 myself, I'm afraid, due to other urgent stuff, but maybe I will eventually. Would be great if someone else had time though.

**Comment by curiousdannii** [comment]
UMD is the current default if EXPORT_ES6 isn't specified. A 'none' option would just be slightly less code, and shouldn't be combined with MODULARIZE/FACTORY (because there'd be no way to access the factory function.) However currently the only way to wrap it all in an IIFE is with MODULARIZE. That would be why renaming would be clearer - it would separate concerns and communicate more clearly about them.

**Comment by sbc100** [comment]
Can you explain the use case for "wrap it all in an IIFE"?    If its not a supper common use case could this IFFE be injected with `--extern-pre-js` and `--extern-post-js`?

**Comment by curiousdannii** [comment]
The main reason would be that if you load it in a regular web page all the variables won't litter global scope. I'd say it's a very common pattern, and it's only 16 characters. I can't think of any reason not to include it for any non ESM output.

**Comment by sbc100** [comment]
I see, that makes sense.  In that case I wonder if we can/should try to make that the default behaviour?  Are there any circumstances where one would want to clutter up the top level namespace?

**Comment by curiousdannii** [comment]
There might be someone out there who accesses internal variables or functions directly instead of exporting them. But that's such an anti-pattern I don't think it should be a concern to us. And if needed they could use `--extern-post-js`.

**Comment by dsyer** [comment]
Here's another workaround.

The hack by @surma (above) worked in Node.js but not in the browser. I ended up building a non-ES6 package with Emscripten, and then using a shim to import it into my application. Starting with this:

```
$ emcc -Os -s EXPORTED_FUNCTIONS="['_doSomething','_main']" bar.c -o bar.js
```

I have `foo.js` (boilerplate) that loads `bar.js` (the generated code)

```js
async function readString(path) {
  if (typeof fetch !== "undefined") {
    const response = await fetch(path);
    const file = await response.text();
    return file;
  }
  const fs = await import('fs');
  return fs.readFileSync(path, "utf8")
}

let preinit = "";
if (typeof fetch === 'undefined') {
  let dirname = (await import("path")).dirname;
  globalThis.__dirname = dirname(import.meta.url);
  let createRequire = (await import('module')).createRequire;
  globalThis.require = createRequire(import.meta.url);
  preinit = "var Module = { locateFile: function(path) { return './' + path; }};\n";
}

var script = await readString('bar.js');
(1, eval)(preinit + script);

async function wait(ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}
let init = async function () {
  while (!Module.calledRun) {
    await wait(500);
  }
}
await init();

export default Module;
```

In the browser:

```html
<html>

<body>
	<script type="importmap">
			{
				"imports": {
					"foo": "./foo.js"
				}
			}
		</script>
	<script type="module">
		import bar from 'foo';
		bar._doSomething();
	</script>
</body>

</html>
```

and in Node.js:

```
$ node
Welcome to Node.js v16.13.1.
Type ".help" for more information.
> var bar = (await import("./foo.js")).default
Starting...
undefined
> bar._doSomething()
```

Others more expert than me in JavaScript can probably improve on that, but it works at least for my simple example. You could remove the hardcoded `foo.js` for instance by using a naming convention based on `import.meta.url`.

**Comment by frank-pian** [comment]
Is there any good advice on React?

**Comment by nick-thompson** [comment]
Any update on this? I appreciate the workarounds, they've been working for me locally but I don't see a good way to extend this generically to users of my library given that some users will run in web and some will run in node. I'd basically have to offload these workarounds onto them, as best as I can see. Would prefer not to do that!

**Comment by curiousdannii** [comment]
You can see here the various environment features Emscripten uses: https://github.com/emscripten-core/emscripten/issues/12203#issuecomment-692387264

**Comment by nick-thompson** [comment]
@curiousdannii thanks, I'm curious in that case if it's possible to set the right set of compile flags to build a result which doesn't reference `path`, `__dirname`, or `fs` at all? I have `-s FILESYSTEM=0` but that doesn't seem to do the trick.

**Comment by curiousdannii** [comment]
`-sENVIRONMENT=web` will exclude all Node inclusions. But if you want to use the one JS output in both browser and Node you'll have to manually load the wasm binary.

**Comment by konsumer** [comment]
What I really want is an ES6 module default export function that lets me slam the bytes in there, and doesn't try to manage getting the wasm bytes. Web & node work the same, if you already have the bytes.

For web I could do it like this:

```js
import mymod from './mymod.js'

const MyMod = mymod(await fetch('mymod.wasm').then(r => r.arrayBuffer()))
```

In node I could do it like this:

```js
import { readFile } from 'fs/promises
const MyMod = mymod(await readFile('mymod.wasm'))
```

for both, I could embed the bytes in the JS file, and it would work great:

```js
const wasmBytes = new Uint8Array([
 // bytes from a tool ike xxd
])

export default mymod(wasmBytes)
```

**Comment by RReverser** [comment]
> Web & node work the same, if you already have the bytes.

They really don't; the generated JS bindings are quite different, especially for things like filesystem support but also others.

**Comment by curiousdannii** [comment]
They *can* work, depending on what you do with it. I use WEB for node in my project.

**Comment by konsumer** [comment]
> They really don't; the generated JS bindings are quite different, especially for things like filesystem support but also others.

I don't mean the existing generated bindings, I mean improving them (as I show in those 2 examples, they work the same.)

I have a [concrete example](https://github.com/konsumer/chiptune2.js). The generated js wrapper had all this stuff to download bytes if on web or read a file in node. My solution was to just gut all that, and [inline the bytes in the file](https://raw.githubusercontent.com/konsumer/chiptune2.js/master/libopenmpt.mjs). Now, it will work on the web, with any web location that doesn't need to be managed, or in a node native AudioContext (polyfilled in C.) if the entry-point itself could take some bytes and just instantiate that, i could manage it however I want (with fetch, read a file, or even just inline the bytes somewhere, and pass them.

**Comment by curiousdannii** [comment]
@konsumer You can pass in the file as `wasmBinary` to the Module constructor. Is that enough for your needs?

**Comment by konsumer** [comment]
I am doing that in the linked js. I would prefer if I had an option to not include all the (dead) code that can "get the bytes".

**Comment by konsumer** [comment]
Maybe I am missing something, but I don't see a way to say "generate all the emscripten wrapping and type converters and stuff, but don't worry about the bytes, I can give that to the module, myself".

**Comment by curiousdannii** [comment]
You might like to subscribe to #12184, it seems to be what you're after.

**Comment by konsumer** [comment]
@curiousdannii I appreciate the link, that does seem to be on at least a similar track. I think maybe a difference is that I want a plain ES6-exported function, that doesn't rely on injecting globals or whatever, that just exports what you need to use the wasm-blob, but you hand it the byte-blob. I'd be happy to make a template for that, I think it's more of a reduction from what there already is, not really an addition.

**Comment by curiousdannii** [comment]
> that doesn't rely on injecting globals

I'm not really sure what you're referring to there, sorry. Nothing in that issue is talking about injecting globals?

**Comment by konsumer** [comment]
I just mean, again, if you look at the linked js, at the very end I do 

```js
export {writeAsciiToMemory, UTF8ToString }
```

so it's possible to actually use the library.

Otherwise it's relying on the fact that ES5 on web has a default `window` global-scope. ES6 doesn;t work like this (and neither does node commonjs) so it won't work like that, but exporting them does.

**Comment by curiousdannii** [comment]
You should be listing those in `EXPORTED_RUNTIME_METHODS` so then they'll be exposed on the Module.

**Comment by kleisauke** [comment]
This will be fixed with PR #17915, see https://github.com/emscripten-core/emscripten/pull/17915#issuecomment-1310758878 for the emsdk install instructions to test this.

**Comment by curiousdannii** [comment]
I don't think this should be closed yet as I don't think we+node builds will work.

**Comment by RReverser** [comment]
> as I don't think web+node builds will work when the output is put through a bundler

Did you try / run into specific issues? As far as I can tell from my usage, it does work, but you need to tweak the bundler config obviously to either ignore or polyfill Node modules  (ignoring is a better option to save code size, but either one works).

But then, you have to do that either way.

**Comment by RReverser** [comment]
> But for it to continue working a bundler would have to be able to output `require()`s when in ESM output mode, and I've never heard of one that does.

Hm, why would you want `require` in your output? Are we talking about bundling - as in, bundling Node.js for the browser?

**Comment by curiousdannii** [comment]
> Hm, why would you want `require` in your output? Are we both talking about bundling from Node.js code for the browser?

Because that's how #17915 works. It will work if you just use the output as is, but if you try to bundle it the bundler will convert the require calls into imports and then it won't work in both web and node.

**Comment by curiousdannii** [comment]
> Could this help? https://github.com/rollup/plugins/tree/master/packages/commonjs#transformmixedesmodules
> 
> I agree that ideally the JS file should still workin in Node and the browser after being bundled by Rollup, Webpack, etc.

Oh, interesting. So yes, there is at least one bundler that could handle it.

**Comment by roj1512** [comment]
Yes, dynamic `import()` should be used.

![image](https://github.com/emscripten-core/emscripten/assets/49933115/2834f594-aadc-4a26-96e0-3152dc5c8928)

**Comment by dasa** [comment]
Could the new Node API `process.getBuiltinModule(id)` help with this by removing the need to use `require` or `import('module')` to load the "fs" and "path" modules.

Doc: https://nodejs.org/docs/latest-v20.x/api/process.html#processgetbuiltinmoduleid

Maybe the code could use this when the min Node version is 20.16 or greater.
